lib/source/consul.js

```diff
@@ -0,0 +1,292 @@
+class Consul extends EventEmitter {
+  /**
+   * Creates a new instance of a Consul souce plugin
+   *
+   * The following options are valid:
+   * - host (String, default 127.0.0.1): Consul agent's address
+   * - port (Number, default 8500) Consul agent's HTTP(S) port
+   * - secure (Boolean, default: true): enable HTTPS
+   *
+   * @param {Object} options  Options that can be set for the plugin
+   */
+  constructor(options) {
+    super();
+
+    const configuration = options || Object.create(null);
+
+    this.type = 'consul';
+    this.name = 'consul';
+    this.properties = Object.create(null);
+    this.host = configuration.host || DEFAULT_AGENT_ADDRESS;
+    this.port = configuration.port || DEFAULT_AGENT_PORT;
+    this.secure = configuration.secure !== false;
+    this._okay = false;
+  }
+
+  /**
+   * Start watching the Consul agent for changes
+   */
+  initialize() {
+    if (this._consul) {
+      return;
+    }
+
+    this._consul = consul({
+      host: this.host,
+      port: this.port,
+      secure: this.secure
+    });
+
+    this._serviceWatcher = this._consul.watch({
+      method: this._consul.catalog.service.list
+    });
+
+    this._serviceWatcher.on('error', (error) => {
+      this._error(error);
+    });
+
+    this._serviceWatcher.on('change', (serviceListData) => {
+      this._onServiceListChange(serviceListData);
+    });
+
+    this._okay = true;
+    this.emit('startup');
+  }
+
+  /**
+   * Stop watching the Consul agent for changes
+   */
+  shutdown() {
+    this._shutdownHealthWatchers();
+    this._shutdownServiceWatcher();
+    delete this._consul;
+
+    this._okay = false;
+    this.emit('shutdown');
+  }
+
+  /**
+   * Return an object describing the Consul plugin's current status
+   *
+   * @return {Object}
+   */
+  status() {
+    return {
+      okay: this._okay,
+      running: this._running
+    };
+  }
+
+  /**
+   * Check the status of the Consul watcher
+   *
+   * @return {Boolean} True if the pooling loop is running
+   * @private
+   */
+  get _running() {
+    return !!this._consul;
+  }
+
+  /* Process updates from Consul's service/list API.
+   *
+   * The data returned Consul looks like this:
+   * [
+   *   "consul": [],
+   *   "service": ["production", "develop"]
+   * ]
+   *
+   * @param {Object} data  The JSON data from the service/list API.
+   * @private
+   */
+  _onServiceListChange(data) {
+    const services = Object.keys(data);
+    const watcherInfo = [];
+
+    services.forEach((service) => {
+      const tags = data[service];
+
+      if (tags.length > 0) {
+        tags.forEach((tag) => {
+          watcherInfo.push({
+            name: service + '-' + tag,
+            options: {
+              passing: true,
+              service,
+              tag
+            }
+          });
+        });
+      } else {
+        watcherInfo.push({
+          name: service,
+          options: {
+            passing: true,
+            service
+          }
+        });
+      }
```

Comment:
This can be simplified because `Array#forEach` won't execute its callback if the array is empty.

```javascript
tags.forEach((tag) => {
  watcherInfo.push({
    name: `${service}-${tag}`,
    options: {
      passing: true,
      service,
      tag
    }
  });
});

if (tags.length === 0) {
  watcherInfo.push({
    name: service,
    options: {
      passing: true,
      service
    }
  });
}
```